web/js/codeworld_shared.js

```diff
@@ -122,13 +122,111 @@ var hintBlacklist = [
     "Pi",
 ];
 
-var codeWorldBuiltinsDocs = {};
+
+// codeWorldDocs is variable containing annotations and documentation
+// of builtin and user-defined variables.
+// Expected format:
+// codeWorldDocs = {
+//   codeWorldLogo: {
+//     annotation: "Picture",
+//     doc: "The CodeWorld logo."
+//   }
+// }
+
+var codeWorldDocs = {};
+
+function renderHover(keyword) {
+    var elem = document.createElement('div')
+
+    if (!codeWorldDocs[keyword]){
+        return null;
+    };
+    elem.title = keyword;
+    var keywordData = codeWorldDocs[keyword];
+
+    // Variable name + annotaion
+    var topLine = document.createElement("div");
+    topLine.className = "hover-decl";
+    var wordElem = document.createElement("span");
+    wordElem.className = "hint-word";
+    wordElem.appendChild(document.createTextNode(keyword));
+    topLine.appendChild(wordElem);
+    if (keywordData.annotation) {
+        var annotation = document.createElement('div');
+        topLine.appendChild(document.createTextNode(" :: " + keywordData.annotation))
+    };
+    elem.appendChild(topLine)
+
+    // Docs
+    if (keywordData.doc) {
+        var docElem = document.createElement('div');
+        docElem.innerHTML = keywordData.doc;
+        docElem.className = "hover-doc";
+        elem.appendChild(docElem)
+    };
+    return elem;
+};
+
+function applyCodemirrorChange(line, change) {
+    if (change.origin === "+input"){
+        return line.slice(0, change.from.ch) + change.text.join('') + line.slice(change.from.ch);
+    } else if (change.origin === "+delete"){
+        return line.slice(0, change.from.ch) + line.slice(change.from.ch + 1);
+    } else if (["setValue", "complete"].indexOf(change.origin) != -1){
+        return;
+    } else {
+        console.log("Event \"" + change.origin + "\" not handled.");
+        return;
+    }
+}
+
+function getWordAndAnnotation(line) {
+    var annotation = null,
+        word = null;
+    // there is top-level assignment
+    if (/^\S+\s*=/.test(line)) {
```

Comment:
Will this match lines like `f(x, y) =`?  (Note the space after the comma.)  I don't think so, since the `=` doesn't immediately follow the first whitespace range.

Reply:
Thanks for correction, will extend parsing rules.